{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notebook for creating CIViC TR submissions to ClinVar. \n",
    "\n",
    "Assumes you have a `.env` file where `CLINVAR_API_KEY` contains the ClinVar Submission API Key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import logging\n",
    "import os\n",
    "import sys\n",
    "from pathlib import Path\n",
    "\n",
    "import requests\n",
    "\n",
    "module_path = os.path.abspath(os.path.join(\"../\"))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from utils import dry_run_test_api  # noqa: E402\n",
    "\n",
    "logging.basicConfig(filename=\"gks-clinvar-civic.log\", level=logging.DEBUG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn = \"civic_studies_20240402.json\"\n",
    "civic_cdm_s3_url = f\"https://raw.githubusercontent.com/ga4gh/gk-pilot/main/datasets/civic/dereferenced/{fn}\"\n",
    "response = requests.get(civic_cdm_s3_url, timeout=5)\n",
    "response.raise_for_status()\n",
    "civic_gks_data = response.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_civic_combo_studies = []\n",
    "all_civic_ta_studies = []\n",
    "\n",
    "for civic_gks in civic_gks_data:\n",
    "  if civic_gks[\"therapeutic\"][\"type\"] == \"CombinationTherapy\":\n",
    "    all_civic_combo_studies.append(civic_gks)\n",
    "  elif civic_gks[\"therapeutic\"][\"type\"] == \"TherapeuticAgent\":\n",
    "    all_civic_ta_studies.append(civic_gks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(99, 610)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_civic_combo_studies), len(all_civic_ta_studies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_condition_set_and_concept_id(study: dict, study_id: str) -> tuple[dict, str] | None:\n",
    "  \"\"\"Get normalized tumor type data from extensions\n",
    "\n",
    "  :param study: GKS study\n",
    "  :param study_id: ID for ``study``\n",
    "  :return: Get condition set data and normalized concept ID if found. Otherwise, ``None``\n",
    "  \"\"\"\n",
    "  tumor_type_exts = study[\"tumorType\"][\"extensions\"]\n",
    "  mondo_id = None\n",
    "  tumor_type_normalized_id = None\n",
    "  for ext in tumor_type_exts:\n",
    "    if ext[\"name\"] == \"disease_normalizer_data\":\n",
    "      ext_val = ext[\"value\"]\n",
    "      mondo_id = ext_val[\"mondo_id\"]\n",
    "      tumor_type_normalized_id = ext_val[\"normalized_id\"]\n",
    "      break\n",
    "\n",
    "  if not mondo_id:\n",
    "    logging.debug(\"%s: no mondo_id\", study_id)\n",
    "    return None\n",
    "\n",
    "  if not tumor_type_normalized_id:\n",
    "    logging.debug(\"%s: no tumor_type_normalized_id\", study_id)\n",
    "    return None\n",
    "\n",
    "  condition_set = {\n",
    "    \"condition\": [{\n",
    "      \"db\": \"MONDO\",\n",
    "      \"id\": mondo_id\n",
    "    }]\n",
    "  }\n",
    "\n",
    "  return condition_set, tumor_type_normalized_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _is_within_indication(reg_approval: dict, tumor_type_normalized_id: str) -> bool:\n",
    "  \"\"\"Determine if drug is FDA approved within indication\n",
    "\n",
    "  :param reg_approval: therapy-normalizer extension for regulatory approval\n",
    "  :param tumor_type_normalized_id: Normalized tumor type concept ID from disease-normalizer\n",
    "  :return: ``True`` if drug is FDA approved within indication. Otherwise, ``False``\n",
    "  \"\"\"\n",
    "  within_indication = False\n",
    "  has_indications = reg_approval.get(\"has_indications\") or []\n",
    "  for ind in has_indications:\n",
    "    mappings = ind.get(\"mappings\") or []\n",
    "    for mapping in mappings:\n",
    "      coding = mapping.get(\"coding\")\n",
    "      if coding and f\"{coding['system']}:{coding['code']}\" == tumor_type_normalized_id:\n",
    "        within_indication = True\n",
    "        break\n",
    "  return within_indication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_ta_ind_and_drug_label(therapeutic: dict, tumor_type_normalized_id: str, study_id: str) -> tuple[bool | None, str | None]:\n",
    "  \"\"\"Get indication and drug label information for therapeutic agent\n",
    "\n",
    "  :param therapeutic: Therapeutic data\n",
    "  :param tumor_type_normalized_id: Normalized tumor type concept ID from disease-normalizer\n",
    "  :param study_id: ID of study\n",
    "  :return: Tuple containing indication and drug label information. Or tuple containing\n",
    "    ``None`` values if no regulatory approval or drug label information is found\n",
    "  \"\"\"\n",
    "  drug_for_ta = None\n",
    "  reg_approval = None\n",
    "  therapy_exts = therapeutic[\"extensions\"]\n",
    "\n",
    "  for ext in therapy_exts:\n",
    "    if ext[\"name\"] == \"therapy_normalizer_data\":\n",
    "      drug_for_ta = ext[\"value\"][\"label\"]\n",
    "    elif ext[\"name\"] == \"regulatory_approval\":\n",
    "      reg_approval = ext[\"value\"]\n",
    "\n",
    "  if not drug_for_ta:\n",
    "    logging.debug(\"%s: no drug_for_ta\", study_id)\n",
    "    return None, None\n",
    "\n",
    "  if not reg_approval:\n",
    "    logging.debug(\"%s: no regulatory_approval\", study_id)\n",
    "    return None, None\n",
    "\n",
    "  within_indication = _is_within_indication(reg_approval, tumor_type_normalized_id)\n",
    "\n",
    "  return within_indication, drug_for_ta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_ind_and_drug_label(study: dict, study_id: str, tumor_type_normalized_id: str) -> tuple[bool, str] | None:\n",
    "  \"\"\"Get indication and drug label information for therapeutic procedure\n",
    "\n",
    "  :param study: GKS study\n",
    "  :param study_id: ID for ``study``\n",
    "  :param tumor_type_normalized_id: Normalized tumor type concept ID from disease-normalizer\n",
    "  :return: Tuple containing indication and drug label if all components found. Otherwise,\n",
    "    ``None``\n",
    "  \"\"\"\n",
    "  therapeutic = study[\"therapeutic\"]\n",
    "  drug_for_ta = None\n",
    "  within_indication = None\n",
    "\n",
    "  if therapeutic[\"type\"] == \"TherapeuticAgent\":\n",
    "    within_indication, drug_for_ta = _get_ta_ind_and_drug_label(\n",
    "      therapeutic, tumor_type_normalized_id, study_id\n",
    "    )\n",
    "  elif therapeutic[\"type\"] == \"CombinationTherapy\":\n",
    "    drug_for_ta_list = []\n",
    "    within_indication_list = []\n",
    "\n",
    "    for comp in therapeutic[\"components\"]:\n",
    "      _within_indication, _drug_for_ta = _get_ta_ind_and_drug_label(\n",
    "        comp, tumor_type_normalized_id, study_id\n",
    "      )\n",
    "      if not _drug_for_ta:\n",
    "        return None\n",
    "\n",
    "      drug_for_ta_list.append(_drug_for_ta)\n",
    "      within_indication_list.append(_within_indication)\n",
    "\n",
    "    drug_for_ta = \";\".join(drug_for_ta_list)\n",
    "    within_indication = all(within_indication_list)\n",
    "\n",
    "  if not drug_for_ta:\n",
    "    return None\n",
    "\n",
    "  return within_indication, drug_for_ta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_clinical_impact_class_descr(vicc_concept_label: str, within_indication: bool, study_id: str) -> str:\n",
    "  \"\"\"Get clinical impact description given vicc concept label\n",
    "\n",
    "  :param vicc_concept_label: VICC concept label\n",
    "  :param within_indication: Whether drug is FDA approved within indication\n",
    "  :param study_id: ID of GKS study\n",
    "  :return: ClinVar clinical impact description\n",
    "  \"\"\"\n",
    "  clinical_impact_class_desc = None\n",
    "  if vicc_concept_label in {\"authoritative evidence\", \"FDA recognized evidence\"}:\n",
    "    if within_indication:\n",
    "      clinical_impact_class_desc = \"Tier I - Strong\"\n",
    "    else:\n",
    "      clinical_impact_class_desc = \"Tier II - Potential\"\n",
    "  elif vicc_concept_label in {\"professional guideline evidence\", \"clinical cohort evidence\"}:\n",
    "    clinical_impact_class_desc = \"Tier I - Strong\"\n",
    "  elif vicc_concept_label in {\"case study evidence\", \"observational study evidence\", \"interventional study evidence\", \"preclinical evidence\"}:\n",
    "    clinical_impact_class_desc = \"Tier II - Potential\"\n",
    "  else:\n",
    "    logging.debug(\"%s: no clinical_impact_class_desc\", study_id)\n",
    "  return clinical_impact_class_desc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_citations(study: dict, study_id: str) -> list[dict]:\n",
    "  \"\"\"Get citation data\n",
    "\n",
    "  :param study: GKS study\n",
    "  :param study_id: ID for ``study``\n",
    "  :return: List of citation data if all components found. Otherwise, empty list\n",
    "  \"\"\"\n",
    "  citations = []\n",
    "  for doc in study[\"isReportedIn\"]:\n",
    "    if doc.get(\"pmid\") and doc[\"pmid\"] is not None:\n",
    "      citations.append({\n",
    "        \"db\": \"PubMed\", \"id\": str(doc[\"pmid\"])\n",
    "      })\n",
    "  if not citations:\n",
    "    logging.debug(\"%s: no citations\", study_id)\n",
    "    return []\n",
    "  return citations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_assertion_type_for_clinical_impact(study: dict, study_id: str) -> str | None:\n",
    "  \"\"\"Get assertion type for clinical impact\n",
    "\n",
    "  :param study: GKS study\n",
    "  :param study_id: ID for ``study``\n",
    "  :return: Assertion type for clinical impact data if all components found. Otherwise,\n",
    "    ``None``\n",
    "  \"\"\"\n",
    "  assertion_type_for_clinical_impact = None\n",
    "  if study[\"predicate\"] == \"predictsSensitivityTo\":\n",
    "    assertion_type_for_clinical_impact = \"therapeutic: sensitivity/response\"\n",
    "  elif study[\"predicate\"] == \"predictsResistanceTo\":\n",
    "    assertion_type_for_clinical_impact = \"therapeutic: resistance\"\n",
    "  else:\n",
    "    logging.debug(\"%s: no assertion_type_for_clinical_impact\", study_id)\n",
    "\n",
    "  return assertion_type_for_clinical_impact\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_variant_set(study: dict, study_id: str) -> dict:\n",
    "  \"\"\"Get variant set data\n",
    "\n",
    "  :param study: GKS study\n",
    "  :param study_id: ID for ``study``\n",
    "  :return: Variant set data if all components found. Otherwise, empty dict\n",
    "  \"\"\"\n",
    "  hgvs_c = None\n",
    "  for expr in study[\"variant\"][\"definingContext\"].get(\"expressions\") or []:\n",
    "    if expr[\"syntax\"] == \"hgvs.c\":\n",
    "      hgvs_c = expr[\"value\"]\n",
    "  if not hgvs_c:\n",
    "    logging.debug(\"%s: no hgvs_c\", study_id)\n",
    "    return {}\n",
    "\n",
    "  ncbi_gene = None\n",
    "  for mapping in study[\"qualifiers\"][\"geneContext\"][\"mappings\"]:\n",
    "    source, code = mapping[\"coding\"][\"code\"].split(\":\")\n",
    "    if source == \"ncbigene\":\n",
    "      ncbi_gene = int(code)\n",
    "  if not ncbi_gene:\n",
    "    logging.debug(\"%s: no ncbi_gene\", study_id)\n",
    "    return {}\n",
    "\n",
    "  return {\n",
    "    \"variant\": [{\n",
    "      \"hgvs\": hgvs_c,\n",
    "      \"gene\": [{\"id\": ncbi_gene}]\n",
    "    }]\n",
    "  }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_clinvar_sub(study: dict) -> dict | None:\n",
    "  \"\"\"Create a ClinVar clinical impact submission for CIViC data\n",
    "\n",
    "  :param study: CIViC study in GKS format\n",
    "  :return: ClinVar impact submission if able to create from study data. Otherwise,\n",
    "    ``None``\n",
    "  \"\"\"\n",
    "  study_id = study[\"id\"]\n",
    "  clinvar_impact_sub = {\n",
    "    \"recordStatus\": \"novel\",\n",
    "    \"localID\": study[\"variant\"][\"id\"],\n",
    "    \"localKey\": study_id,\n",
    "    \"observedIn\": [{\n",
    "      \"alleleOrigin\": \"somatic\",\n",
    "      \"affectedStatus\": \"yes\",\n",
    "      \"collectionMethod\": \"curation\"\n",
    "    }]\n",
    "  }\n",
    "\n",
    "  # conditionSet\n",
    "  condition_set_data = _get_condition_set_and_concept_id(study, study_id)\n",
    "  if not condition_set_data:\n",
    "    return None\n",
    "  condition_set, tumor_type_normalized_id = condition_set_data\n",
    "  clinvar_impact_sub[\"conditionSet\"] = condition_set\n",
    "\n",
    "  # drugForTherapeuticAssertion\n",
    "  ind_and_drug_label_data = _get_ind_and_drug_label(study, study_id, tumor_type_normalized_id)\n",
    "  if not ind_and_drug_label_data:\n",
    "    return None\n",
    "  within_indication, drug_for_ta = ind_and_drug_label_data\n",
    "\n",
    "  # clinicalImpactClassificationDescription\n",
    "  vicc_concept_label = study.get(\"strength\").get(\"label\")\n",
    "  clinical_impact_class_desc = _get_clinical_impact_class_descr(\n",
    "    vicc_concept_label, within_indication, study_id\n",
    "  )\n",
    "  if not clinical_impact_class_desc:\n",
    "    return None\n",
    "\n",
    "  # citations\n",
    "  citations = _get_citations(study, study_id)\n",
    "  if not citations:\n",
    "    return None\n",
    "\n",
    "  # assertionTypeForClinicalImpact\n",
    "  assertion_type_for_clinical_impact = _get_assertion_type_for_clinical_impact(study, study_id)\n",
    "  if not assertion_type_for_clinical_impact:\n",
    "    return None\n",
    "\n",
    "  # variantSet\n",
    "  variant_set = _get_variant_set(study, study_id)\n",
    "  if not variant_set:\n",
    "    return None\n",
    "  clinvar_impact_sub[\"variantSet\"] = variant_set\n",
    "\n",
    "  clinvar_impact_sub[\"clinicalImpactClassification\"] = {\n",
    "    \"clinicalImpactClassificationDescription\": clinical_impact_class_desc,\n",
    "    \"assertionTypeForClinicalImpact\": assertion_type_for_clinical_impact,\n",
    "    \"drugForTherapeuticAssertion\": drug_for_ta,\n",
    "    \"comment\": study[\"description\"],\n",
    "    \"citation\": citations\n",
    "  }\n",
    "\n",
    "  return clinvar_impact_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_civic_ta_studies = []\n",
    "valid_civic_combo_studies = []\n",
    "\n",
    "civic_eid_2997 = []\n",
    "combo_sub = []\n",
    "batch_sub_mp12 = []\n",
    "mpids = set()\n",
    "batch_sub = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_tr_sub(studies: list[dict], valid_studies: list[dict], is_combo: bool) -> None:\n",
    "  \"\"\"Add TR submissions to associated lists\n",
    "\n",
    "  :param studies: Studies for therapeutic agent or combination therapy\n",
    "  :param valid_studies: List of valid studies for therapeutic agent or combination\n",
    "    therapy. This will be mutated\n",
    "  \"\"\"\n",
    "  cutoff = 5 if is_combo else 2\n",
    "  for study in studies:\n",
    "    sub = create_clinvar_sub(study)\n",
    "    if sub:\n",
    "      valid_studies.append(sub)\n",
    "\n",
    "      if sub[\"localKey\"] == \"civic.eid:2997\":\n",
    "        civic_eid_2997.append(sub)\n",
    "        continue\n",
    "\n",
    "      if is_combo and not combo_sub:\n",
    "        combo_sub.append(sub)\n",
    "      else:\n",
    "        if len(batch_sub_mp12) < cutoff and sub[\"localID\"] == \"civic.mpid:12\":\n",
    "          batch_sub_mp12.append(sub)\n",
    "        else:\n",
    "          if len(batch_sub) < cutoff and sub[\"localID\"] not in mpids:\n",
    "            mpids.add(sub[\"localID\"])\n",
    "            batch_sub.append(sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(71, 434)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "add_tr_sub(all_civic_ta_studies, valid_civic_ta_studies, False)\n",
    "add_tr_sub(all_civic_combo_studies, valid_civic_combo_studies, True)\n",
    "len(valid_civic_combo_studies), len(valid_civic_ta_studies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "for fn, sub_data in [\n",
    "  (\"therapeutic_evidence.json\", civic_eid_2997),\n",
    "  (\"combination_therapy_evidence.json\", combo_sub),\n",
    "  (\"batch_mp12_therapy_evidence.json\", batch_sub_mp12),\n",
    "  (\"batch_therapeutic_evidence.json\", batch_sub)\n",
    "]:\n",
    "  clin_impact_sub = {\"clinicalImpactSubmission\": sub_data}\n",
    "  response = dry_run_test_api(clin_impact_sub)\n",
    "\n",
    "  if response.status_code == 204:\n",
    "    with Path(fn).open(\"w\") as wf:\n",
    "      json.dump(clin_impact_sub, wf, indent=2)\n",
    "      wf.write(\"\\n\")\n",
    "  else:\n",
    "    msg = f\"{fn} is not valid: {response.json()}\"\n",
    "    raise Exception(msg)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
